src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/SentenceCombinationService.kt

```diff
@@ -0,0 +1,161 @@
+      }
+    }
+
+    // delete all the concurrentSentencesStream sentences
+    for (concurrentSentence in concurrentSentences) {
+      offenderSentenceProfile.getSentences().remove(concurrentSentence)
+    }
+    val earliestSentencedAt: Unit = concurrentSentences.stream()
+      .map(Sentence::getSentencedAt)
+      .max(Comparator.nullsFirst(Comparator.comparing(LocalDate::toEpochDay)))
+    if (earliestSentencedAt.isPresent()) {
+      val sentence = Sentence(
+        concurrentSentences.get(0).getOffence(),
+        earliestSentencedAt.get(),
+        combinedDuration,
+        offenderSentenceProfile
+      )
+      sentence.calculate()
+      // question here about how are we supposed to deal the CRD/ARD
+      sentence.getSentenceCalculation().setIsReleaseDateConditional(isReleaseDateConditional)
+      // add a new sentence with the new duration
+      offenderSentenceProfile.addSentence(sentence)
+    }
+
+    return workingOffenderSentenceProfile
+  }
+
+   */
+
+  private fun combineSentenceWithEachOfItsConcurrentSentences(
+    sentence: Sentence,
+    workingOffenderSentenceProfile: OffenderSentenceProfile
+  ) {
+    val concurrentSentences = sentence.concurrentSentences.toList()
+    for (concurrentSentence in concurrentSentences) {
+      combineTwoSentences(sentence, concurrentSentence, workingOffenderSentenceProfile)
+    }
+  }
+
+  fun combineTwoSentences(
+    firstSentence: Sentence,
+    secondSentence: Sentence,
+    workingOffenderSentenceProfile: OffenderSentenceProfile
+  ) {
+
+    // I take 2 sentences, combine them into a single conjoined sentence
+    val combinedSentence: Sentence = mergeSentences(firstSentence, secondSentence)
+    combinedSentence.associateSentences(mutableListOf())
+
+    addSentenceToProfile(combinedSentence, workingOffenderSentenceProfile)
+
+    associateExistingLinksToNewSentence(firstSentence, combinedSentence, workingOffenderSentenceProfile)
+    associateExistingLinksToNewSentence(secondSentence, combinedSentence, workingOffenderSentenceProfile)
+
+    // delete the 2 existing sentences
+    deleteSentenceFromProfile(firstSentence, workingOffenderSentenceProfile)
+    deleteSentenceFromProfile(secondSentence, workingOffenderSentenceProfile)
+
+    removeSelfFromConcurrentSentences(combinedSentence)
+  }
+
+  private fun removeSelfFromConcurrentSentences(sentence: Sentence) {
+    for (UUID in sentence.concurrentSentenceUUIDs) {
+      if (UUID == sentence.identifier) {
+        sentence.concurrentSentenceUUIDs.remove(UUID)
+      }
+    }
+    for (sentenceObject in sentence.concurrentSentences) {
+      if (sentenceObject == sentence) {
+        sentence.concurrentSentences.remove(sentenceObject)
+      }
+    }
+  }
+
+  private fun associateExistingLinksToNewSentence(
+    originalSentence: Sentence,
+    replacementSentence: Sentence,
+    offenderSentenceProfile: OffenderSentenceProfile
+  ) {
+
+    for (sentence in offenderSentenceProfile.sentences) {
```

Comment:
same - could use a more functional way of looping rather the for loops. Happy  to defer